modules/text2vec-weaviate/clients/weaviate_embed.go

```diff
@@ -0,0 +1,223 @@
+		return nil, errors.Errorf(errorMessage)
+	}
+
+	var resBody embeddingsResponse
+	if err := json.Unmarshal(bodyBytes, &resBody); err != nil {
+		return nil, errors.Wrap(err, "unmarshal response body")
+	}
+
+	if len(resBody.Embeddings) == 0 {
+		return nil, errors.Errorf("empty embeddings response")
+	}
+
+	return &modulecomponents.VectorizationResult{
+		Text:       input,
+		Dimensions: len(resBody.Embeddings[0]),
+		Vector:     resBody.Embeddings,
+	}, nil
+}
+
+func (v *vectorizer) getWeaviateEmbedURL(ctx context.Context, baseURL string) string {
+	passedBaseURL := baseURL
+	if headerBaseURL := modulecomponents.GetValueFromContext(ctx, "X-Weaviate-Baseurl"); headerBaseURL != "" {
+		passedBaseURL = headerBaseURL
+	}
+	return v.urlBuilder.url(passedBaseURL)
+}
+
+func (v *vectorizer) getEmbeddingsRequest(texts []string, isSearchQuery bool, dimensions *int64) embeddingsRequest {
+	return embeddingsRequest{Texts: texts, IsSearchQuery: isSearchQuery, Dimensions: dimensions}
+}
+
+func (v *vectorizer) GetApiKeyHash(ctx context.Context, config moduletools.ClassConfig) [32]byte {
+	key, err := v.getApiKey(ctx)
+	if err != nil {
+		return [32]byte{}
+	}
+	return sha256.Sum256([]byte(key))
+}
+
+func (v *vectorizer) GetVectorizerRateLimit(ctx context.Context, cfg moduletools.ClassConfig) *modulecomponents.RateLimits {
+	rpm, _ := modulecomponents.GetRateLimitFromContext(ctx, "Weaviate", DefaultRPM, 0)
+
+	execAfterRequestFunction := func(limits *modulecomponents.RateLimits, tokensUsed int, deductRequest bool) {
+		// refresh is after 60 seconds but leave a bit of room for errors. Otherwise, we only deduct the request that just happened
+		if limits.LastOverwrite.Add(61 * time.Second).After(time.Now()) {
+			if deductRequest {
+				limits.RemainingRequests--
+			}
+			return
+		}
+
+		limits.RemainingRequests = rpm
+		limits.ResetRequests = time.Now().Add(time.Duration(61) * time.Second)
+		limits.LimitRequests = rpm
+		limits.LastOverwrite = time.Now()
+
+		// high dummy values for tokens per minute
+		limits.RemainingTokens = DefaultTPM
+		limits.LimitTokens = DefaultTPM
+		limits.ResetTokens = time.Now().Add(time.Duration(1) * time.Second)
+	}
+
+	initialRL := &modulecomponents.RateLimits{AfterRequestFunction: execAfterRequestFunction, LastOverwrite: time.Now().Add(-61 * time.Minute)}
+	initialRL.ResetAfterRequestFunction(0) // set initial values
+
+	return initialRL
+}
+
+func getErrorMessage(statusCode int, resBodyError string, errorTemplate string) string {
+	var errResp embeddingsResponseError
+	if err := json.Unmarshal([]byte(resBodyError), &errResp); err != nil {
+		return fmt.Sprintf(errorTemplate, statusCode, resBodyError)
+	}
+	return fmt.Sprintf(errorTemplate, statusCode, errResp.Detail)
+}
+
+func (v *vectorizer) getApiKey(ctx context.Context) (string, error) {
+	if apiKey := modulecomponents.GetValueFromContext(ctx, "X-Weaviate-Embedding-Key"); apiKey != "" {
```

Comment:
maybe we should simplify the headers and use `X-Weaviate-Api-Key` naming?

Reply:
also please whitelist your header [here](https://github.com/weaviate/weaviate/blob/main/usecases/config/config_handler.go#L299)